### PR TITLE
Adds icepack BGC 

### PR DESCRIPTION
**Pull request by njeffery**
Updates the version of icepack to one that includes sea ice BGC

Tested in ocean-sea ice BGC active simulations.

[NML]
[non-BFB] only for tests with biogeochemistry and/or sea ice aerosols active

### CODE REVIEW

**Review by eclare108213 — COMMENTED** [review]
Initial comments:  I'd like to run some D tests on the code.  Here are some things I noticed, just reading through the code diffs.  Comments on mpas-seaice code are direct.  For Icepack:

- Where is this new submodule hash located (which fork/branch)? Is it different from the submodule listed in E3SM/.gitsubmodules?
- Do the changes in icepack_aerosol.F90 change the answers for ‘old’ aerosol runs? e.g. is subroutine update_snow_bgc called when tr_aero=T?
- icepack_parameters.F90:  Is grid_o_t used in icepack?  I had removed it from the parameters list because it wasn't used.
- icepack_therm_itd.F90: The first initialization of vicen_init (originally =0, now =vicen(:)) in lateral_melt could be removed completely, since it’s set later.  Or initialize both vicen_init and vsnon_init here and remove the later ones.
- icedrv_init_column.F90: There’s a new cpp, NEWINITZBGC.  Should that be here?  I don't think it's set anywhere in the icepack driver.

FYI @apcraig

File: components/mpas-seaice/src/shared/mpas_seaice_icepack.F [hunk 1]
```diff
@@ -15237,65 +14455,38 @@ subroutine init_column_biogeochemistry_profiles(domain, tracerObject)
...
+                  fed=oceanDissolvedIronConc(:,iCell), &
+                  fep=oceanParticulateIronConc(:,iCell), &
+                  zaeros=oceanZAerosolConc(:,iCell), &
+                  ocean_bio_all=oceanBioConcentrations(:,iCell), &
+                  hum=oceanHumicsConc(iCell))
+
+             call icepack_init_bgc(&
+                  ncat=nCategories, &
+                  nblyr=nBioLayers, &
+                  nilyr=nIceLayers, &
```
[the remark this answers is not in the document]
Comment (apcraig) [reply components/mpas-seaice/src/shared/mpas_seaice_icepack.F:14479]: This is part of the cleanup I was hoping to do with the bgc update.  But first priority was to converge on the init_zbgc.  I still would like to go thru the bgc interfaces carefully and identify things that could be removed and things that should be made optional.

**Comment by apcraig** [comment]
> Initial comments: I'd like to run some D tests on the code. Here are some things I noticed, just reading through the code diffs. Comments on mpas-seaice code are direct. For Icepack:
> 
>     * Where is this new submodule hash located (which fork/branch)? Is it different from the submodule listed in E3SM/.gitsubmodules?
> 
>     * Do the changes in icepack_aerosol.F90 change the answers for ‘old’ aerosol runs? e.g. is subroutine update_snow_bgc called when tr_aero=T?
> 
>     * icepack_parameters.F90:  Is grid_o_t used in icepack?  I had removed it from the parameters list because it wasn't used.
> 
>     * icepack_therm_itd.F90: The first initialization of vicen_init (originally =0, now =vicen(:)) in lateral_melt could be removed completely, since it’s set later.  Or initialize both vicen_init and vsnon_init here and remove the later ones.
> 
>     * icedrv_init_column.F90: There’s a new cpp, NEWINITZBGC.  Should that be here?  I don't think it's set anywhere in the icepack driver.
> 
> 
> FYI @apcraig

I'm testing the new icepack now too.  The NEWINITZBGC was something I put in there as part of the transition/recommendation to show where I was hoping we would go.  I am going to have a PR to NJ's Icepack branch to clean this up.  

Also, I checked for grid_o_t.  It's not used in Icepack and no longer appears in the Icepack driver or CICE.

**Comment by apcraig** [comment]
@njeffery, has any testing been done with "debug" build flags.  I'm running into several divide by zeros and other things are part of my testing of standalone Icepack.  I will try to implement code fixes and then create a PR to your Icepack branch.  Then we can discuss whether some of the fixes are due to incorrect setup of a case or other things.

**Comment by njeffery** [comment]
> @njeffery, has any testing been done with "debug" build flags. I'm running into several divide by zeros and other things are part of my testing of standalone Icepack. I will try to implement code fixes and then create a PR to your Icepack branch. Then we can discuss whether some of the fixes are due to incorrect setup of a case or other things.

Yes.  I've compiled with debug.  I'll test again was I've added some of Elizabeth's corrections.

**Comment by njeffery** [comment]
@eclare108213 and @apcraig :  The corresponding icepack branch is E3SM-Project/Icepack.git njeffery/fixes-to-bgc

**Comment by njeffery** [comment]
> Initial comments: I'd like to run some D tests on the code. Here are some things I noticed, just reading through the code diffs. Comments on mpas-seaice code are direct. For Icepack:
> 
>     * Where is this new submodule hash located (which fork/branch)? Is it different from the submodule listed in E3SM/.gitsubmodules?

E3SM-Project/Icepack.git njeffery/fixes-to-bgc 
> 
>     * Do the changes in icepack_aerosol.F90 change the answers for ‘old’ aerosol runs? e.g. is subroutine update_snow_bgc called when tr_aero=T?

None of these changes should impact the old aerosols
> 
>     * icepack_parameters.F90:  Is grid_o_t used in icepack?  I had removed it from the parameters list because it wasn't used.

This was originally for zsalinity.  I believe it can be removed.

>     * icepack_therm_itd.F90: The first initialization of vicen_init (originally =0, now =vicen(:)) in lateral_melt could be removed completely, since it’s set later.  Or initialize both vicen_init and vsnon_init here and remove the later ones.
> 
>     * icedrv_init_column.F90: There’s a new cpp, NEWINITZBGC.  Should that be here?  I don't think it's set anywhere in the icepack driver.
> 
> 
> FYI @apcraig

**Comment by proteanplanet** [comment]
Hello @njeffery.  Looking through the array of testing being undertaken, I might best help by running B-cases in E3SM.  I've never run BGC in a B-case before, but the fastest realistic test platform would be the standard V3 resolution configuration with BGC.  Do you know if this is set up?  I will await Tony's commits and Elizabeth's D-case checks before I start testing, but make preparations for this B-case testing in the meantime.

**Comment by njeffery** [comment]
@proteanplanet :  I'd like to see a fully coupled bgc run, but I'm not sure that's the best comparison.  Kat's been running a pre-Industrial bgc run with master, and it just crashed at year 39 with a MarBL error.  Although the MarBL port has been well tested in GCASEs, until now, it hasn't been tested in full coupling.  So, we should try a short run. Maybe 20 years?

**Comment by apcraig** [comment]
> > @njeffery, has any testing been done with "debug" build flags. I'm running into several divide by zeros and other things are part of my testing of standalone Icepack. I will try to implement code fixes and then create a PR to your Icepack branch. Then we can discuss whether some of the fixes are due to incorrect setup of a case or other things.
> 
> Yes. I've compiled with debug. I'll test again was I've added some of Elizabeth's corrections.

Derecho has been down the last two days for emergency maintenance.  Hoping it's back up again today, so I can continue to test.  I've run into both divide by zeros and array bounds issues in the standalone testing.  I don't know if something about one (or all) of the four columns we test is unusual.  Also, I'm testing by a bgc and an skl bgc case.  You are running with the debug flags with the bgc on?  Are we still supporting skl?

**Comment by apcraig** [comment]
> @eclare108213 and @apcraig : The corresponding icepack branch is E3SM-Project/Icepack.git njeffery/fixes-to-bgc

@njeffery, Are you working in the E3SM-Project repo or your repo?  Just want to make sure I'm using the latest code and want to make sure I PR any suggested changes to the right place.  It would be helpful to move forward on a specific branch in a specific repo and stay there.  I don't care where that is.  Thanks.

**Comment by njeffery** [comment]
E3SM-Project

**Comment by proteanplanet** [comment]
> @proteanplanet : I'd like to see a fully coupled bgc run, but I'm not sure that's the best comparison. Kat's been running a pre-Industrial bgc run with master, and it just crashed at year 39 with a MarBL error. Although the MarBL port has been well tested in GCASEs, until now, it hasn't been tested in full coupling. So, we should try a short run. Maybe 20 years?

OK. Let me digest this an think about what test we would run in a B-case to demonstrate there are no problems in the full icepack.  We need a 100 year simulation for PRs of this kind to be accepted.

**Comment by njeffery** [comment]
I'd like to try a 100 year zaerosol fully coupled test.  I did a long one shortly before Dec when trying to get dust/BC in v3. It doesn't use MarBL so shouldn't crash. It won't replace a bgc test, but works as an additional test.

**Comment by njeffery** [comment]
> > > @njeffery, has any testing been done with "debug" build flags. I'm running into several divide by zeros and other things are part of my testing of standalone Icepack. I will try to implement code fixes and then create a PR to your Icepack branch. Then we can discuss whether some of the fixes are due to incorrect setup of a case or other things.
> > 
> > 
> > Yes. I've compiled with debug. I'll test again was I've added some of Elizabeth's corrections.
> 
> Derecho has been down the last two days for emergency maintenance. Hoping it's back up again today, so I can continue to test. I've run into both divide by zeros and array bounds issues in the standalone testing. I don't know if something about one (or all) of the four columns we test is unusual. Also, I'm testing by a bgc and an skl bgc case. You are running with the debug flags with the bgc on? Are we still supporting skl?

I haven't tested skl and don't have plans to test it in E3SM.  To my knowledge, it's never been tested in MPAS.

**Comment by apcraig** [comment]
> > > > @njeffery, has any testing been done with "debug" build flags. I'm running into several divide by zeros and other things are part of my testing of standalone Icepack. I will try to implement code fixes and then create a PR to your Icepack branch. Then we can discuss whether some of the fixes are due to incorrect setup of a case or other things.
> > > 
> > > 
> > > Yes. I've compiled with debug. I'll test again was I've added some of Elizabeth's corrections.
> > 
> > 
> > Derecho has been down the last two days for emergency maintenance. Hoping it's back up again today, so I can continue to test. I've run into both divide by zeros and array bounds issues in the standalone testing. I don't know if something about one (or all) of the four columns we test is unusual. Also, I'm testing by a bgc and an skl bgc case. You are running with the debug flags with the bgc on? Are we still supporting skl?
> 
> I haven't tested skl and don't have plans to test it in E3SM. To my knowledge, it's never been tested in MPAS.

Do we expect skl to work in general and will we continue to support it in the Consortium?  If we are, then it's important that it work robustly.  For the Consortium, we need to make sure we're clear what bgc configurations are supported and to test all those configurations.   What E3SM uses in production is a related by somewhat separate issue.

**Comment by apcraig** [comment]
I have created a PR with some changes to Icepack for the Icepack standalone bgc runs.  There are still problems that I'm working on but it's at least a start.  https://github.com/E3SM-Project/Icepack/pull/37

I have only tested this version of the Icepack model.  I will next
- review all changes relative to the Consortium main
- test relative to the Consortium main Icepack
- test in CICE

I am still getting a bgc skl array bounds error in this version of the model with build debug flags on that I need to better understand and may need some help with.

**Comment by apcraig** [comment]
There is some new code in icepack_biogeochemistry (see below).   This is where the skl debug tests are hitting an array bounds error in this term, 

`trcrn(bio_index(mm) + k-1,n).`

Any thoughts about how this should be fixed?

```
forrtl: severe (408): fort: (2): Subscript #1 of the array TRCRN has value 39 which is greater than the upper bound of 38
Image              PC                Routine            Line        Source             
icepack            0000000000972446  icepack_zbgc_mp_i         945  icepack_zbgc.F90
icepack            00000000005641A9  icedrv_step_mp_bi        1442  icedrv_step.F90
```

```
+         else
+            do mm = 1, nbtrcr
+               do k  = 1, nblyr+1
+                  if (present(flux_bion)) then
+                     flux_bion(mm,n) = flux_bion(mm,n) + trcrn(bio_index(mm) + k-1,n) *  &
+                        hin_old(n) * zspace(k)/dt * trcrn(nt_fbri,n)
+                  endif
+                  flux_bio(mm) = flux_bio(mm) + trcrn(bio_index(mm) + k-1,n) * &
+                     vicen_init(n) * zspace(k)/dt * trcrn(nt_fbri,n)
+                  trcrn(bio_index(mm) + k-1,n) = c0
+                enddo
+            enddo
```

**Comment by apcraig** [comment]
It looks like cgrid, bgrid, igrid, icgrid, and swgrid are all definied at initialization, identical on each column, and fixed throughout the run.  These are computed in icepack then passed back to the driver which then never uses the values, but passes them back and forth into icepack.  Could these "grid" variables be initialized via a call to icepack_init_hbrine then just stored in Icepack permanently.  Just looking at ways to clean up the bgc interfaces.

**Comment by njeffery** [comment]
@eclare108213 and @apcraig : Two 15-year GMPAS tests completed, one with icepack enabled and the other with column_package.  See https://acme-climate.atlassian.net/wiki/spaces/ICE/pages/4405002400/dust.icepack.GMPAS.IcoswISC30E3r5.BGCfinal+and+dust.column.GMPAS.IcoswISC30E3r5.BGCall   

The carbon conservation member in the column_package version had some spurious high values which appeared now and again.  This was traced to errors in the bgc implementation for step_therm2.  The Icepack version no longer has these high values.

The sea ice primary production monthly climatology averaged over the last 5 years of the run looks identical in log plots.

I'll merge @apcraig's updates to icepack and let you know when I've verified BFB.

**Comment by njeffery** [comment]
@eclare108213  and @apcraig :  Merged Tony's changes to Icepack and updated the mpas interface.  Also verified that the code is BFB.

**Comment by apcraig** [comment]
@njeffery, where are the Icepack commit in the E3SM-Project fork.  I don't see them on [njeffery/fixes-to-bgc](https://github.com/E3SM-Project/Icepack/tree/njeffery/fixes-to-bgc) or anywhere else, am I missing something?

**Comment by njeffery** [comment]
@proteanplanet :  Would you like me to set up a couple of short runs to verify BFB with BGC off?

**Comment by njeffery** [comment]
@jonbob :  This PR has a new stream "DustIronMonthlyForcing".  It reads a modified version of the ocean ecoForcingMonthly file.  So far, the only version I know of is here: /lcrc/group/e3sm/data/inputdata/ocn/mpas-o/IcoswISC30E3r5/ecoForcingSurfaceMonthly+WetDryDustSolFrac.IcoswISC30E3r5.20240511.nc

I added lines 590-597 to components/mpas-seaice/cime_config/buildnml  to add this in, but they should be generalized.  Could you take a look and let me know how they should be changed.

The file is used only if config_use_zaerosols  = true, and either or both of the following are true config_use_iron_solubility_file and config_use_atm_dust_file.

**Comment by proteanplanet** [comment]
@njeffery I will do it with the ice test script.

**Comment by njeffery** [comment]
Thanks @proteanplanet .

**Comment by jonbob** [comment]
@njeffery -- do you envision a new file with the "DustIronMonthlyForcing" stream for other meshes as well? I'm assuming it is mesh specific

**Comment by njeffery** [comment]
@jonbob :  Yes.  it'll be the same stream used in the ocean model when BGC is on  "ecosys_monthly_flux".

**Comment by eclare108213** [comment]
> I am still getting a bgc skl array bounds error in this version of the model with build debug flags on ... Any thoughts about how this should be fixed?

@njeffery pointed out that the problem code has a loop over nblyr, but skl should only be a single layer.  E3SM doesn't run skl and so we'll need to decide what to do about it in the Consortium, independent of E3SM.

**Comment by jonbob** [comment]
@njeffery -- I pushed a couple of commits back to your branch:
1.  Update the bld files to match Registry -- very minor but I did add back the logic for which column physics to use, as we discussed
2. added support for the file related to the new DustIronMonthlyForcing stream be mesh-dependent and more generalized

I tested the DustIron part in a couple of cases, one where it showed up in the streams.seaice as expected and one where it did not, also expected. But please feel free to rename the variable I added. I also made sure the new file is on the inputdata repo

**Review by njeffery — COMMENTED** [review]
Great.  Thanks @jonbob.

**Review by njeffery — COMMENTED** [review]
Perfect.

**Comment by apcraig** [comment]
Is the Icepack version that goes with this commit still https://github.com/E3SM-Project/Icepack/tree/njeffery/fixes-to-bgc?  It looks like there are no updates to that branch in the last 3 weeks, is that correct?  Also, just FYI, this E3SM PR looks to me like it's not pointing to that version of Icepack.  I understand that update may still need to be done and/or there may be other E3SM processes to do so, just wanted to point it out.  Thanks.

**Comment by njeffery** [comment]
@apcraig :  It's pointing to #31ce422.

**Comment by apcraig** [comment]
@njeffery, I see it now, sorry about my confusion.  I needed to "see more commits" on the Icepack diffs link.  Thanks.  Looks good.

**Comment by njeffery** [comment]
@apcraig :  Whew!  I had a little panic there for a moment...

**Comment by proteanplanet** [comment]
Can I proceed with the BFB B-case BGC/No-BGC test now?  I will be merging this branch into a branch from the current E3SM master, and then running a 12-month test of both with BGC switched off.

**Comment by njeffery** [comment]
Yes.  Thanks @proteanplanet

**Comment by rljacob** [comment]
status:  still getting tested with long comparisons.

**Comment by njeffery** [comment]
@proteanplanet :  Let me know if you have any concerns with this PR.  It would be great if we could get this through this week.  Thanks.

**Comment by proteanplanet** [comment]
B-case tests are in progress on Chrysalis.

**Review by proteanplanet — APPROVED** [review]
Additions and changes in this branch were cherry picked and tested in a branch from master on July 29, 2024 with BGC switched off. This test was to confirm that when sea ice BGC is switched off and brought into the code as a stealth feature, default B-case simulations using master are unaffected.  Two parallel 12-month B-cases were integrated at standard V3 resolution - one forked from master current to [466fa81](https://github.com/proteanplanet/E3SM/commit/466fa81cd251af23d82739e252eb3042ca684bed), the second using the same fork, but with all commits particular to this branch out to [cf6f75b](https://github.com/proteanplanet/E3SM/commit/cf6f75b2bae91f770d37eddbfdbff0830b22f420).  The results are that sea ice is unaffected across 10 metrics, indicating the model climate as whole is unaffected:
```
    iceAreaCell 	BFB
    iceVolumeCell 	BFB
    icePressure 	BFB
    uVelocityGeo 	BFB
    vVelocityGeo 	BFB

    totalIceExtent 	BFB
    totalIceVolume 	BFB
    totalSnowVolume 	BFB
    totalKineticEnergy 	BFB
    averageAlbedo 	BFB
```
Therefore I approve this PR.

**Comment by njeffery** [comment]
@proteanplanet : Great!  thank you!

**Comment by jonbob** [comment]
@eclare108213 -- can you sign off on this?

**Review by eclare108213 — APPROVED** [review]
Approved based on extensive testing by @njeffery and @proteanplanet.  @apcraig is also testing the icepack changes in CICE, where problems appear to be related to configuration choices rather than to the BGC code itself.  If we turn up changes needed in Icepack, we will submit them separately.  Kudos to @njeffery -- this has been a HUGE amount of work!

**Comment by jonbob** [comment]
passes e3sm_ice_developer suite on chrysalis with expected NML DIFFs

**Comment by njeffery** [comment]
Yay!   Thank you Everyone!

**Comment by jonbob** [comment]
merged to next

**Comment by jonbob** [comment]
merged to master and expected DIFFs blessed -- except for one test that is still running on pm-cpu

**Pull request merged**